### PR TITLE
`Development`: Modernize appearance of charts on programming exercise grading page

### PR DESCRIPTION
**Pull request by willergeorg**
#### General

- [ ] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.ase.in.tum.de/dev/guidelines/language-guidelines/).
- [ ] I chose a title conforming to the [naming conventions for pull requests](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/development-process.html#naming-conventions-for-github-pull-requests).
#### Client
- [x] I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client/).
- [ ] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client-tests/).
- [x] I documented the TypeScript code using JSDoc style.
- [x] I added multiple screenshots/screencasts of my UI changes.
- [x] I translated all newly inserted strings into English and German.

### Motivation and Context

This PR is changing the underlying framework for the charts located in programming exercise grading from `ng2-charts`/`chart.js `to `ngx-charts`.

### Description

The html templates as well as the underlying logic in the components are refactored in order to fit the requests of `ngx-charts`.

### Steps for Testing

Prerequisites:
- 1 Instructor
- 1 student
- 1 Programming Exercise with static code analysis enabled
- It might be easier to have two windows with the users in parallel rather than logging out and logging in again. This can be realized via different browsers or using incognito mode.

**Important**: Before deploying this PR, take a look at the current charts. Especially check the tooltips and that the colors in the list on the left and in the charts are mapped correctly (e.g. light blue bar represents value of the list entry associated with the color light blue). Taking screenshots might ease the comparison of the different designs.

1. Log in to Artemis as instructor
2. Navigate to Course Administration
3. Click on "Exercises" of the corresponding course
4. In the row of the programming exercise click on the "Grading" button
5. Navigate to the tab "Code Analysis"
6. In the list on the right, feel free to play with the configuration parameters. Inspect if the chart still changes as with the old implementation. (afaik with no submissions, only the first bar should change)
7. **Don't forget to save your changes**
8. Go back to the "Test cases" tab. Again, be creative with the different parameters and make sure that the Weight distribution chart still works as before. ("Total points" should stay empty if nobody participated on the exercise)
9.  **Don't forget to save your changes**
10. Switch to the student account **but if possible keep the grading page open for the instructor**
11. Participate on the exercise and submit your participation. Try to trigger the static code analysis e.g. by inserting unused public attributes.
12. After the submission is build, as instructor switch to the grading page again and go to the "Code Analysis" tab (maybe refreshing might be necessary but not sure). Make sure that the chart is updated correctly.
13. Go back to the "Test Cases" tab. If your submission achieved any points, "Total points" should be updated appropriately. 
14. For all of those charts, the tooltips  should be translated now (which was not the case before). Check this as well.
#### Code Review
- [x] Review 1
- [x] Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

### Screenshots

Old design of Category Distribution:
image

*New* Design of Category Distribution:
image

Old Design of Weight Distribution:
image

*New* Design of Weight Distribution:
image

Old Design of Total Points:
image

*New* Design of Total Points:
image

### CODE REVIEW

File: src/main/webapp/app/exercises/programming/manage/grading/charts/test-case-distribution-chart.component.ts [hunk 1]
```diff
@@ -74,50 +143,73 @@ export class TestCaseDistributionChartComponent implements OnChanges {
...
-                    backgroundColor,
-                    hoverBackgroundColor,
-                    data: [element.relPoints],
-                });
+                const color = this.getColor(+i / this.testCases.length, 50);
+
+                weight.series.push({ name: element.label, value: element.relWeight, bonus: element.relScore });
+                weightAndBonus.series.push({ name: element.label, value: element.relScore, weight: element.relWeight });
+
+                points.series.push({ name: element.label, value: element.relPoints });
```
Comment (b-fein) [inline src/main/webapp/app/exercises/programming/manage/grading/charts/test-case-distribution-chart.component.ts:169]:
```suggestion
                weight.series.push({ name: label, value: element.relWeight, bonus: element.relScore });
                weightAndBonus.series.push({ name: label, value: element.relScore, weight: element.relWeight });

                points.series.push({ name: label, value: element.relPoints });
```
`const label = element.label` in line 163

File: src/main/webapp/app/exercises/programming/manage/grading/charts/test-case-distribution-chart.component.ts [hunk 2]
```diff
@@ -74,50 +143,73 @@ export class TestCaseDistributionChartComponent implements OnChanges {
...
     }
+
+    /**
+     * Appends a percentage sign to every tick on the x axis
+     * @param tick the default tick label as string
+     * @returns tick label string that is extended by an percentage sign
+     */
+    xAxisFormatting(tick: string): string {
+        return tick + '%';
+    }
```
Comment (b-fein) [inline src/main/webapp/app/exercises/programming/manage/grading/charts/test-case-distribution-chart.component.ts:214]: This is already identically in `sca-category-distribution-chart.component.ts`. Can it be moved to a utility file to reduce duplication?

File: src/main/webapp/i18n/de/programmingAssessment.json [hunk 3]
```diff
@@ -5,7 +5,13 @@
                 "heading": "Einreichungen und Bewertungen der Textaufgabe \"{{exerciseTitle}}\""
             },
             "confirmCancel": "Bist du sicher, dass du die Bewertung abbrechen möchtest? Dein aktueller Bewertungsfortschritt wird gelöscht und die Abgabe steht anderen Tutor:innen wieder zur Verfügung.",
-            "notFound": "Wir haben keine neue Programmier Abgabe ohne Bewertung gefunden, bitte gehe zurück."
+            "notFound": "Wir haben keine neue Programmier Abgabe ohne Bewertung gefunden, bitte gehe zurück.",
+            "penalty": "Penalty",
+            "issues": "Issues",
```
Comment (b-fein) [inline src/main/webapp/i18n/de/programmingAssessment.json:10]: IIRC in other places those non-translations are used as well. ‘Punktabzug’ and ‘Probleme’ would probably be better.

(Just a note so I don’t forget about it, should probably be changed consistently everywhere in a separate PR.)

**Review by b-fein — COMMENTED** [review]
Manual tests: During the testing session the charts worked as expected (some issues were found if the input in the data table were invalid ⇒ #4484, out of scope here).

Code review: I added a few inline comments, only small remarks.

File: src/main/webapp/app/exercises/programming/manage/grading/charts/test-case-distribution-chart.component.ts [hunk 4]
```diff
@@ -74,50 +143,73 @@ export class TestCaseDistributionChartComponent implements OnChanges {
...
-                    data: [element.relWeight, element.relScore],
-                });
-
-                this.pointsChartDatasets.push({
-                    label,
-                    backgroundColor,
-                    hoverBackgroundColor,
-                    data: [element.relPoints],
-                });
+                const color = this.getColor(+i / this.testCases.length, 50);
```
[the remark this answers is not in the document]
Comment (willergeorg) [reply src/main/webapp/app/exercises/programming/manage/grading/charts/test-case-distribution-chart.component.ts]: Changed in f29991a

File: src/main/webapp/i18n/de/programmingAssessment.json [hunk 5]
```diff
@@ -5,7 +5,13 @@
                 "heading": "Einreichungen und Bewertungen der Textaufgabe \"{{exerciseTitle}}\""
             },
             "confirmCancel": "Bist du sicher, dass du die Bewertung abbrechen möchtest? Dein aktueller Bewertungsfortschritt wird gelöscht und die Abgabe steht anderen Tutor:innen wieder zur Verfügung.",
-            "notFound": "Wir haben keine neue Programmier Abgabe ohne Bewertung gefunden, bitte gehe zurück."
+            "notFound": "Wir haben keine neue Programmier Abgabe ohne Bewertung gefunden, bitte gehe zurück.",
```
[the remark this answers is not in the document]
Comment (willergeorg) [reply src/main/webapp/i18n/de/programmingAssessment.json]: f29991a

File: src/main/webapp/app/exercises/programming/manage/grading/charts/sca-category-distribution-chart.component.ts [hunk 6]
```diff
@@ -1,20 +1,68 @@
...
+                    <ng-template #tooltipTemplate let-model="model">
+                        <b>{{ model.name }}</b>
+                        <br />
+                        <div [ngSwitch]="model.series">
+                            <div *ngSwitchCase="'Penalty'">
+                                <span>{{ 'artemisApp.programmingAssessment.penaltyTooltip' | artemisTranslate: { percentage: model.value.toFixed(2) } }}</span>
+                                <br />
+                                <span>
+                                    {{ 'artemisApp.programmingAssessment.issuesTooltip' | artemisTranslate: { percentage: model.issues.toFixed(2) } }}
+                                </span>
```
[the remark this answers is not in the document]
Comment (willergeorg) [reply src/main/webapp/app/exercises/programming/manage/grading/charts/sca-category-distribution-chart.component.ts]: Changed in f29991a

File: src/main/webapp/app/exercises/programming/manage/grading/charts/test-case-distribution-chart.component.ts (hunk 1 above)
Comment (willergeorg) [reply src/main/webapp/app/exercises/programming/manage/grading/charts/test-case-distribution-chart.component.ts]: Changed in f29991a

File: src/main/webapp/app/exercises/programming/manage/grading/charts/test-case-distribution-chart.component.ts (hunk 2 above)
Comment (willergeorg) [reply src/main/webapp/app/exercises/programming/manage/grading/charts/test-case-distribution-chart.component.ts]: I introduced a utils in f29991a

File: src/main/webapp/i18n/de/programmingAssessment.json (hunk 3 above)
Comment (willergeorg) [reply src/main/webapp/i18n/de/programmingAssessment.json:10]: I agree that a translation would be better. I decided to stick with the English terms to keep it consistent to the explanation text above the chart. There, the terms stay in English as well.

**Review by b-fein — APPROVED** [review]
Thanks for addressing my comments. Code looks good now. :+1:

**Review by ole-ve — CHANGES_REQUESTED** [review]
Tested on TS1 and charts update accordingly.
I found an issue when negative values are inserted for
- Test Cases Grading for any cell
Bildschirmfoto 2021-12-10 um 11 41 24
- SCA Grading for any cell when grading is active for the corresponding category
Bildschirmfoto 2021-12-10 um 11 39 35

I tested and confirm that this was not a problem on the live server. As validation is done on saving, I suggest only fixing the chart appearance and not making any validation on the input fields.

**Review by Malyuk-A — APPROVED** [review]
Tested on TS1. Works as described.
I discussed the code with @willergeorg and I think it is reasonable.

**Review by ole-ve — APPROVED** [review]
Tested on TS1, works as described.
Charts are also robust against negative values.

**Review by b-fein — APPROVED** [review]
Code re-approve. :+1:

**Review by filip-gregurevic — APPROVED** [review]
Looks good to me 👍🏼
